### PR TITLE
Update the operator documentation generation

### PR DESCRIPTION
**Pull request by skottmckay**
**Description**: 
Update the operator documentation generation
  - Add ability to filter what's included in the doc generation so the CI build can limit the doc updates to just the things we care about
    - Include CPU and CUDA kernels
  - Move doc generation to GPU build so CUDA operator kernels are included
  - Add ability for CI to publish updated docs as a build artifact if there are diffs
    - developer can download generated docs and add to PR to resolve diffs instead of having to do a local build with CUDA enabled
  - Update to latest supported operators
  - Make layout a little nicer
  - Fix incorrect usage of 'onnx.ai.ml' as the default domain
    - ML ops are now separated from the real default domain of 'onnx.ai'

**Motivation and Context**
Fix issues with documentation.

### CODE REVIEW

File: tools/ci_build/github/azure-pipelines/win-gpu-ci-pipeline.yml [hunk 1]
```diff
@@ -147,12 +148,28 @@ jobs:
 
   - powershell: |
      Get-ChildItem -Path dist/*.whl | foreach {pip --disable-pip-version-check install --upgrade $_.fullname}
-     python $(Build.SourcesDirectory)\tools\ci_build\build.py --config $(BuildConfig) --build_dir $(Build.BinariesDirectory) $(UseOmp) --skip_submodule_sync --build_shared_lib --test --cmake_generator "Visual Studio 16 2019" --build_wheel --disable_rtti --use_winml  --build_shared_lib --enable_onnx_tests --enable_wcos $(additionalBuildFlags)
-   
+     python $(Build.SourcesDirectory)\tools\ci_build\build.py --config $(BuildConfig) --build_dir $(Build.BinariesDirectory) $(UseOmp) --test --enable_onnx_tests  --gen_doc validate $(additionalBuildFlags)
```
Comment (pranavsharma) [inline tools/ci_build/github/azure-pipelines/win-gpu-ci-pipeline.yml:151]: Bunch of build flags missing. Intentional?

File: tools/python/gen_opkernel_doc.py [hunk 2]
```diff
@@ -44,36 +43,49 @@ def format_param_strings(params):
...
-        fout.write('## Supported Operators Data Types\n')
+    return providers
+
+
+def main(output_path: pathlib.Path, provider_filter: [str]):
+
+    providers = expand_providers(provider_filter)
+
+    with io.open(output_path, 'w', newline='', encoding="utf-8") as fout:
+        fout.write('## Supported Operators and Data Types\n')
```
Comment (pranavsharma) [inline tools/python/gen_opkernel_doc.py:68]: In OperatorKernels.md it's not clear if other execution providers are also part of the documentation. Is it possible to have TOC hyperlinks?

File: tools/ci_build/build.py [hunk 3]
```diff
@@ -184,9 +183,9 @@ def parse_arguments():
         available test data directories.""")
 
     # generate documentation
-    parser.add_argument(
-        "--gen_doc", action='store_true',
-        help="Generate documentation on contrib ops")
+    parser.add_argument("--gen_doc", nargs='?', const='yes', type=str,
+                        help="Generate operator kernels and contrib operator documentation. "
```
Comment (pranavsharma) [inline tools/ci_build/build.py:187]: nit: Generate documentation listing standard ONNX operators and types implemented by various execution providers and contrib operator schemas.

File: tools/ci_build/github/azure-pipelines/win-gpu-ci-pipeline.yml (hunk 1 above)
Comment (skottmckay) [reply tools/ci_build/github/azure-pipelines/win-gpu-ci-pipeline.yml:151]: Yeah. They're values that aren't used unless `--update` is specified.

File: tools/python/gen_opkernel_doc.py (hunk 2 above)
Comment (skottmckay) [reply tools/python/gen_opkernel_doc.py:68]: The TOC stuff in gh-pages is coming from the site generation infrastructure. I could try and replicate that in the python but that requires some logic to also create anchors. Does it need to be part of this checkin? Short term could we just list the EPs covered at the top and add the anchors/links in a separate PR?

Comment (pranavsharma) [reply tools/python/gen_opkernel_doc.py:68]: Not urgent. Can be different PR.

**Review by pranavsharma — APPROVED** [review]

**Review by pranavsharma — APPROVED** [review]

**Review by pranavsharma — APPROVED** [review]

**Pull request merged**